File: android/domain/src/main/java/com/rasoiai/domain/model/Notification.kt
package com.rasoiai.domain.model

/**
 * Types of notifications in the app.
 */
enum class NotificationType(val value: String) {
    /** Festival reminders for upcoming festivals */
    FESTIVAL_REMINDER("festival_reminder"),

    /** Meal plan updates/regeneration notifications */
    MEAL_PLAN_UPDATE("meal_plan_update"),

    /** Shopping/grocery reminders */
    SHOPPING_REMINDER("shopping_reminder"),

    /** Personalized recipe suggestions */
    RECIPE_SUGGESTION("recipe_suggestion"),

    /** Cooking streak milestone achievements */
    STREAK_MILESTONE("streak_milestone");

    companion object {
        fun fromValue(value: String): NotificationType {
            return entries.find { it.value == value } ?: MEAL_PLAN_UPDATE
        }
    }
}

/**
 * Types of actions that can be performed from a notification.
 */
enum class NotificationActionType(val value: String) {
    /** Open a specific recipe */
    OPEN_RECIPE("open_recipe"),

    /** Open meal plan screen */
    OPEN_MEAL_PLAN("open_meal_plan"),

    /** Open grocery list */
    OPEN_GROCERY("open_grocery"),

    /** Open stats/achievements screen */
    OPEN_STATS("open_stats"),

    /** No action - just informational */
    NONE("none");

    companion object {
        fun fromValue(value: String?): NotificationActionType {
            return entries.find { it.value == value } ?: NONE
        }
    }
}

/**
 * Represents a user notification.
 *
 * @property id Unique identifier
 * @property type Type of notification
 * @property title Notification title
 * @property body Notification body/message
 * @property imageUrl Optional image URL for rich notifications
 * @property actionType Type of action to perform when tapped
 * @property actionData Data needed for the action (e.g., recipeId)
 * @property isRead Whether the notification has been read
 * @property createdAt Timestamp when notification was created
 * @property expiresAt Optional timestamp after which notification should be hidden
 */
data class Notification(
    val id: String,
    val type: NotificationType,
    val title: String,
    val body: String,
    val imageUrl: String? = null,
    val actionType: NotificationActionType = NotificationActionType.NONE,
    val actionData: NotificationActionData? = null,
    val isRead: Boolean = false,
    val createdAt: Long,
    val expiresAt: Long? = null
) {
    /** Check if notification is expired */
    val isExpired: Boolean
        get() = expiresAt != null && expiresAt <= System.currentTimeMillis()

    /** Check if notification has an action */
    val hasAction: Boolean
        get() = actionType != NotificationActionType.NONE
}

/**
 * Data class for notification action payload.
 */
data class NotificationActionData(
    val recipeId: String? = null,
    val mealPlanId: String? = null,
    val festivalId: String? = null,
    val streakCount: Int? = null
)

File: backend/app/models/notification.py
"""Notification database models."""

import uuid
from datetime import datetime
from typing import TYPE_CHECKING, Optional

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base, TimestampMixin

if TYPE_CHECKING:
    from app.models.user import User


class Notification(Base, TimestampMixin):
    """User notification model.

    Types:
    - festival_reminder: Upcoming festival alerts
    - meal_plan_update: Meal plan changes/regeneration
    - shopping_reminder: Grocery shopping reminders
    - recipe_suggestion: Personalized recipe recommendations
    - streak_milestone: Cooking streak achievements
    """

    __tablename__ = "notifications"

    id: Mapped[str] = mapped_column(
        String(36),
        primary_key=True,
        default=lambda: str(uuid.uuid4()),
    )
    user_id: Mapped[str] = mapped_column(
        String(36),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    type: Mapped[str] = mapped_column(String(50), nullable=False, index=True)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    body: Mapped[str] = mapped_column(Text, nullable=False)
    image_url: Mapped[Optional[str]] = mapped_column(String(500), nullable=True)

    # Action configuration
    action_type: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)
    action_data: Mapped[Optional[str]] = mapped_column(Text, nullable=True)  # JSON

    # Status
    is_read: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False, index=True)
    expires_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True), nullable=True)

    # Relationships
    user: Mapped["User"] = relationship("User", back_populates="notifications")


class FcmToken(Base, TimestampMixin):
    """FCM token storage for push notifications."""

    __tablename__ = "fcm_tokens"

    id: Mapped[str] = mapped_column(
        String(36),
        primary_key=True,
        default=lambda: str(uuid.uuid4()),
    )
    user_id: Mapped[str] = mapped_column(
        String(36),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    token: Mapped[str] = mapped_column(String(500), nullable=False, unique=True)
    device_type: Mapped[str] = mapped_column(String(20), nullable=False, default="android")
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)

    # Relationships
    user: Mapped["User"] = relationship("User", back_populates="fcm_tokens")
